desktop/renderer/index.js

```diff
@@ -52,59 +48,34 @@ class Keybase extends Component {
     // Used by material-ui widgets.
     injectTapEventPlugin()
 
-    // For remote window components
-    ipcMain.removeAllListeners('dispatchAction')
-    ipcMain.removeAllListeners('stateChange')
-    ipcMain.removeAllListeners('subscribeStore')
-
-    ipcMain.on('dispatchAction', (event, action) => {
-      // we MUST clone this else we'll run into issues with redux. See https://github.com/rackt/redux/issues/830
-      // This is because we get a remote proxy object, instead of a normal object
-      // Using _.cloneDeep() creates a non-plain object
-      setImmediate(() => store.dispatch(_.merge({}, action)))
-    })
-
-    ipcMain.on('subscribeStore', event => {
-      const sender = event.sender // cache this since this is actually a sync-rpc call...
-
-      // Keep track of the last state sent so we can make the diffs.
-      let oldState = {}
-      const getStore = () => {
-        const newState = store.getState()
-        const diffState = shallowDiff(oldState, newState) || {}
-        oldState = newState
-        return diffState
-      }
+    this.setupDispatchAction()
+    this.setupStoreSubscriptions()
 
-      if (!sender.isDestroyed()) {
-        sender.send('stateChange', getStore())
-      }
+    // Handle notifications from the service
+    store.dispatch(listenForNotifications())
 
-      let unsubscribe = store.subscribe(() => {
-        if (sender.isDestroyed()) {
-          if (unsubscribe) {
-            unsubscribe()
-            unsubscribe = null
-          }
-          return
-        }
+    // Handle logUi.log
+    ListenLogUi()
+  }
 
-        const diffState = getStore()
-        if (Object.keys(diffState).length !== 0) {
-          if (!sender.isDestroyed()) { // We need this extra check due to timing issues
-            sender.send('stateChange', diffState)
```

Comment:
It's asyc from main. We're in tender so we're proxying main. We basically
should have on handlers on ipcmain not in the main thread

On Wednesday, January 20, 2016, Marco Munizaga <notifications@github.com>
wrote:

> In desktop/renderer/index.js
> <https://github.com/keybase/client/pull/1772#discussion_r50301923>:
>
> >
> > -        const diffState = getStore()
> > -        if (Object.keys(diffState).length !== 0) {
> > -          if (!sender.isDestroyed()) { // We need this extra check due to timing issues
> > -            sender.send('stateChange', diffState)
>
> Sounds like a bug in electron because docs say this is an async reply:
> http://electron.atom.io/docs/v0.36.4/api/ipc-main/#sending-messages
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/keybase/client/pull/1772/files#r50301923>.
>